File: survive/game/procgen.py
"""This module handles various procedural generation tasks."""


import collections
import math
import random
from typing import Dict, List, Optional, Set

from .attributes import AttributeSet
from .constants import (CHALLENGE_LEVEL_SCALE_UP_FACTOR,
                        CREATURE_MAX_DAMAGE_AT_LEVEL_1,
                        CREATURE_MAX_HP_AT_LEVEL_1, CREATURE_MIN_HP_AT_LEVEL_1)
from .creature import Creature
from .game import game
from .item import Armor, Weapon
from .types import Wieldpoint
from .words import random_adjective, random_adverb

NAMES: Dict[str, Wieldpoint] = {
    "Helmet": "head",
    "Cap": "head",
    "Chestplate": "chest",
    "Shirt": "chest",
    "Leggings": "legs",
    "Greaves": "legs",
    "Boots": "feet",
    "Shoes": "feet",
    "Sandals": "feet",
    "Gauntlets": "arms",
    "Vambraces": "arms",
    "Sleeves": "arms",
    "Sword": "hands",
    "Dagger": "hands",
    "Stick": "hands",
    "Bat": "hands",
    "Longsword": "hands",
    "Shiv": "hands",
}

MOB_ATTRIBUTES = AttributeSet()
MOB_ATTRIBUTES.modify("str", 16)
MOB_ATTRIBUTES.modify("dex", 14)
MOB_ATTRIBUTES.modify("con", 12)
MOB_ATTRIBUTES.modify("int", 12)
MOB_ATTRIBUTES.modify("wis", 13)
MOB_ATTRIBUTES.modify("chr", 10)


class NameGenerator:
    """A name generator that never uses the same name twice."""

    def __init__(self, rng: random.Random):
        self._seen: Set[str] = set()
        self._rng = rng

        self._point_to_names: collections.defaultdict[
            Wieldpoint, List[str]
        ] = collections.defaultdict(list)

        for name, point in NAMES.items():
            self._point_to_names[point].append(name)

    def generate_name(self, point: Wieldpoint, special=False) -> str:
        """Generate a name for the given wieldpoint."""

        names_list = self._point_to_names[point]

        adj = random_adjective(self._rng)
        if special:
            adverb = random_adverb(self._rng)

        while True:
            name = self._rng.choice(names_list)
            if special:
                name = f"The {adverb} {adj} {name}"
            else:
                name = f"The {adj} {name}"

            if name in self._seen:
                adj = random_adjective(self._rng)
                if special:
                    adverb = random_adverb(self._rng)
                continue

            self._seen.add(name)

            return name


def create_weapon(name: str, max_damage: int, challenge_level: int, rng: random.Random):
    """Create a procedurally-generated weapon."""

    dmg = rng.randint(max_damage // 2, max_damage)

    faces = 20
    if dmg < faces:
        faces = dmg

    count = int(math.floor(dmg / faces))

    bonus = challenge_level

    return Weapon(name.title(), 20, 2, bonus + 1, 0, f"{count}d{faces}")


def create_armor(
    name: str, point: Wieldpoint, challenge_level: int, rng: random.Random
):
    """Create a procedurally-generated armor."""

    return Armor(
        point,
        name=name.title(),
        attackbonus=challenge_level,
        defensebonus=rng.randint(1, challenge_level + 2),
    )


def creature_at_level(
    challenge_level: int,
    name_generator: NameGenerator,
    rng: Optional[random.Random] = None,
) -> Creature:
    """Create a procedurally-generated creature at the given challenge level."""

    if rng is None:
        rng = game().random()

    max_damage = CREATURE_MAX_DAMAGE_AT_LEVEL_1

    min_hp = CREATURE_MIN_HP_AT_LEVEL_1
    max_hp = CREATURE_MAX_HP_AT_LEVEL_1

    potential_mountpoints: List[Wieldpoint] = ["chest", "arms", "feet"]

    for _ in range(challenge_level):
        # scale up
        max_damage = int(math.ceil(max_damage * CHALLENGE_LEVEL_SCALE_UP_FACTOR))
        max_hp = int(math.ceil(max_hp * CHALLENGE_LEVEL_SCALE_UP_FACTOR))
        min_hp = int(math.ceil(min_hp * CHALLENGE_LEVEL_SCALE_UP_FACTOR))

    weapon_name = name_generator.generate_name("hands")
    weapon = create_weapon(weapon_name, max_damage, challenge_level, rng)

    point: Wieldpoint = rng.choice(potential_mountpoints)

    armor_name = name_generator.generate_name(point)
    armor = create_armor(armor_name, point, challenge_level, rng)

    total_hp = rng.randint(min_hp, max_hp)

    creature = Creature(None, (0, 0), "Goblin")
    creature.hitpoints = creature.maxhitpoints = total_hp
    creature.wield(armor.wields_at(), armor)
    creature.wield(weapon.wields_at(), weapon)

    return creature

File: web/src/procgen.ts
import {
  CHALLENGE_LEVEL_SCALE_UP_FACTOR,
  CREATURE_BASE_CON,
  CREATURE_BASE_DEX,
  CREATURE_BASE_DEFENSE_BONUS,
  CREATURE_BASE_ATTACK_BONUS,
  CREATURE_BASELINE_CON,
  CREATURE_BASELINE_DEX,
  CREATURE_BASELINE_STR,
  CREATURE_BASE_STR,
  CREATURE_MAX_DAMAGE_AT_LEVEL_1,
  CREATURE_MAX_HP_AT_LEVEL_1,
  CREATURE_MIN_HP_AT_LEVEL_1,
  ITEM_ARMOR_DEFENSE_BASE_MIN,
  ITEM_ARMOR_DEFENSE_CHALLENGE_BONUS,
  ITEM_WEAPON_FLAT_BONUS_BASE,
  ITEM_WEAPON_FLAT_BONUS_CHALLENGE_DIVISOR,
  ITEM_WEAPON_FLAT_BONUS_RANDOM_MAX,
  NAME_GENERATION_FALLBACK_MAX_ID,
  NAME_GENERATION_MAX_ATTEMPTS,
  PLAYER_CRIT_MAXIMUM_MULTIPLIER,
  PLAYER_CRIT_MINIMUM_MULTIPLIER,
  PLAYER_CRIT_MINIMUM_ROLL,
  WEAPON_CRIT_MAXIMUM_ROLL,
  WEAPON_DAMAGE_FACE_MAX,
  WEAPON_DAMAGE_FACE_MIN,
} from "./constants";
import { AttributeSet } from "./attributes";
import { Armor, Weapon } from "./items";
import { Creature } from "./creature";
import { SeededRandom } from "./rng";
import type { WieldSlot } from "./types";

const NAME_PARTS: Record<WieldSlot, string[]> = {
  head: ["Helmet", "Cap", "Mask"],
  chest: ["Chestplate", "Shirt", "Vest"],
  arms: ["Gauntlets", "Sleeves", "Vambraces"],
  hands: ["Sword", "Dagger", "Club", "Mace", "Spear"],
  legs: ["Greaves", "Leggings", "Pants"],
  feet: ["Boots", "Shoes", "Sandals"],
};

const ADJECTIVES = [
  "ancient",
  "rusted",
  "vicious",
  "sturdy",
  "ornate",
  "ashen",
  "scarred",
  "silent",
  "grim",
  "hollow",
  "tempered",
  "radiant",
];

const ADVERBS = [
  "eerily",
  "violently",
  "surprisingly",
  "quietly",
  "wildly",
  "ominously",
];

export class NameGenerator {
  private readonly seen = new Set<string>();

  constructor(private readonly rng: SeededRandom) {}

  generateName(slot: WieldSlot, special = false): string {
    const options = NAME_PARTS[slot];
    for (
      let attempt = 0;
      attempt < NAME_GENERATION_MAX_ATTEMPTS;
      attempt += 1
    ) {
      const noun = this.rng.choice(options);
      const adjective = this.rng.choice(ADJECTIVES);
      const value = special
        ? `The ${this.rng.choice(ADVERBS)} ${adjective} ${noun}`
        : `The ${adjective} ${noun}`;

      if (this.seen.has(value)) {
        continue;
      }
      this.seen.add(value);
      return value;
    }

    const fallback = `${this.rng.choice(ADJECTIVES)} ${this.rng.choice(options)} ${Math.floor(this.rng.next() * NAME_GENERATION_FALLBACK_MAX_ID)}`;
    this.seen.add(fallback);
    return fallback;
  }
}

export function createWeapon(
  name: string,
  maxDamage: number,
  challengeLevel: number,
  rng: SeededRandom,
): Weapon {
  const damage = rng.int(
    Math.max(1, Math.floor(maxDamage / 2)),
    Math.max(2, maxDamage),
  );
  const faces = Math.max(
    WEAPON_DAMAGE_FACE_MIN,
    Math.min(WEAPON_DAMAGE_FACE_MAX, damage),
  );
  const count = Math.max(1, Math.floor(damage / faces));
  const flatBonus = Math.max(
    ITEM_WEAPON_FLAT_BONUS_BASE,
    ITEM_WEAPON_FLAT_BONUS_BASE +
      Math.floor(challengeLevel / ITEM_WEAPON_FLAT_BONUS_CHALLENGE_DIVISOR) +
      rng.int(0, ITEM_WEAPON_FLAT_BONUS_RANDOM_MAX),
  );

  const critRange = rng.int(PLAYER_CRIT_MINIMUM_ROLL, WEAPON_CRIT_MAXIMUM_ROLL);
  const critMult = rng.int(
    PLAYER_CRIT_MINIMUM_MULTIPLIER,
    PLAYER_CRIT_MAXIMUM_MULTIPLIER,
  );

  return new Weapon(
    name,
    critRange,
    critMult,
    challengeLevel + 1,
    0,
    `${count}d${faces} +${flatBonus}`,
  );
}

export function createArmor(
  name: string,
  slot: WieldSlot,
  challengeLevel: number,
  rng: SeededRandom,
): Armor {
  return new Armor(
    slot,
    name,
    challengeLevel,
    rng.int(
      ITEM_ARMOR_DEFENSE_BASE_MIN,
      challengeLevel + ITEM_ARMOR_DEFENSE_CHALLENGE_BONUS,
    ),
  );
}

export function creatureAtLevel(
  challengeLevel: number,
  names: NameGenerator,
  rng: SeededRandom,
): Creature {
  let maxDamage = CREATURE_MAX_DAMAGE_AT_LEVEL_1;
  let minHp = CREATURE_MIN_HP_AT_LEVEL_1;
  let maxHp = CREATURE_MAX_HP_AT_LEVEL_1;

  for (let i = 0; i < challengeLevel; i += 1) {
    maxDamage = Math.ceil(maxDamage * CHALLENGE_LEVEL_SCALE_UP_FACTOR);
    maxHp = Math.ceil(maxHp * CHALLENGE_LEVEL_SCALE_UP_FACTOR);
    minHp = Math.ceil(minHp * CHALLENGE_LEVEL_SCALE_UP_FACTOR);
  }

  const creature = new Creature(
    "Goblin",
    { x: 0, y: 0 },
    new AttributeSet({
      str: CREATURE_BASELINE_STR,
      dex: CREATURE_BASELINE_DEX,
      con: CREATURE_BASELINE_CON,
      int: 8,
      wis: 8,
      chr: 8,
    }),
  );
  creature.attributes.modify("str", CREATURE_BASE_STR);
  creature.attributes.modify("dex", CREATURE_BASE_DEX);
  creature.attributes.modify("con", CREATURE_BASE_CON);
  creature.attackBonus = CREATURE_BASE_ATTACK_BONUS;
  creature.defenseBonus = CREATURE_BASE_DEFENSE_BONUS;

  const weapon = createWeapon(
    names.generateName("hands"),
    maxDamage,
    challengeLevel,
    rng,
  );
  const armorSlot: WieldSlot = rng.choice(["chest", "arms", "feet"]);
  const armor = createArmor(
    names.generateName(armorSlot),
    armorSlot,
    challengeLevel,
    rng,
  );

  creature.wield("hands", weapon);
  creature.wield(armorSlot, armor);

  creature.maxHitpoints = rng.int(minHp, maxHp);
  creature.hitpoints = creature.maxHitpoints;

  return creature;
}
